src/phoenix/core/model_schema.py

```diff
@@ -0,0 +1,1067 @@
+    URLs. All Dimensions have a primary column and its name is the same as
+    the Dimension's. Some Dimensions may have more than one column, but only
+    one is primary. When a Dimension tries to extract a non-existent column
+    from a dataframe, a default (or fallback) column will be returned, e.g.
+    a column of NaNs.
+    """
+
+    _datasets: Dict[DatasetRole, Dataset]
+    _dimensions: Dict[Name, Dimension]
+    _dim_roles: Dict[DimensionRole, List[Name]]
+    _original_columns: Dict[DatasetRole, pd.Index]
+    # now is used as the substitute for dataframes without timestamps.
+    _now: datetime
+
+    def __init__(
+        self,
+        dimensions: Iterable[Dimension],
+        dataframes: Iterable[Union[pd.DataFrame, Tuple[Name, pd.DataFrame]]],
+        /,
+        treat_omitted_columns_as_features: bool = True,
+        timestamps_already_normalized: bool = False,
+        df_already_sorted_by_time: bool = False,
+        # TODO: Consider moving validations here.
+        df_already_validated: bool = False,
+    ):
+        df_names, dfs = cast(
+            Tuple[Iterable[Name], Iterable[pd.DataFrame]],
+            zip(*_ensure_tuple(dataframes)),
+        )
+        str_col_dfs = _ensure_str_column_names(dfs)
+        padded_dfs = _pad(str_col_dfs, pd.DataFrame)
+        padded_df_names = _pad(df_names, _rand_str)
+        datasets = starmap(
+            self._new_dataset,
+            zip(padded_dfs, padded_df_names, DatasetRole),
+        )
+        # Store datasets by role.
+        object.__setattr__(
+            self,
+            "_datasets",
+            {ds.role: ds for ds in datasets},
+        )
+        # Preserve originals, useful for exporting.
+        object.__setattr__(
+            self,
+            "_original_columns",
+            {role: ds.columns for role, ds in self._datasets.items()},
+        )
+
+        # Store dimensions by name. In general, a dimension's name is that of
+        # its primary column. But dimensions with special roles, e.g.
+        # `PREDICTION_ID` have names that match their roles, e.g.
+        # `Prediction ID` (without underscores), and don't necessarily match
+        # the names of actual underlying columns in the dataframes.
+        object.__setattr__(
+            self,
+            "_dimensions",
+            {dim.name: self._new_dimension(dim) for dim in dimensions},
+        )
+        if treat_omitted_columns_as_features:
+            self._dimensions.update(
+                (name, self._new_dimension(name, role=FEATURE))
+                for name in _get_omitted_column_names(
+                    self._dimensions.values(),
+                    self._datasets.values(),
+                )
+            )
+
+        # Group dimension names by roles. Store in a `defaultdict(list)`.
+        object.__setattr__(
+            self,
+            "_dim_roles",
+            defaultdict(list, _group_names_by_dim_role(self._dimensions.values())),
+        )
```

Comment:
For my context, why is the object prototype function being invoked like this? Why can't we just set it? This is very hard to read